File: lib/dns-stack.ts
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export interface DnsStackArgs {
  environmentSuffix: string;
  domainName?: string;
  albDnsName: pulumi.Output<string>;
  albZoneId: pulumi.Output<string>;
  tags: pulumi.Input<{ [key: string]: string }>;
}

export class DnsStack extends pulumi.ComponentResource {
  public readonly hostedZoneId?: pulumi.Output<string>;
  public readonly nameServers?: pulumi.Output<string[]>;
  public readonly recordName?: pulumi.Output<string>;

  constructor(
    name: string,
    args: DnsStackArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('tap:dns:DnsStack', name, args, opts);

    // Only create DNS resources if domain name is provided
    if (args.domainName) {
      // Create Route53 hosted zone
      const hostedZone = new aws.route53.Zone(
        `payment-zone-${args.environmentSuffix}`,
        {
          name: args.domainName,
          comment: `Hosted zone for payment processing platform - ${args.environmentSuffix}`,
          tags: pulumi.all([args.tags]).apply(([t]) => ({
            ...t,
            Name: `payment-zone-${args.environmentSuffix}`,
          })),
        },
        { parent: this }
      );

      // Create A record pointing to ALB
      const record = new aws.route53.Record(
        `payment-record-${args.environmentSuffix}`,
        {
          zoneId: hostedZone.zoneId,
          name: args.domainName,
          type: 'A',
          aliases: [
            {
              name: args.albDnsName,
              zoneId: args.albZoneId,
              evaluateTargetHealth: true,
            },
          ],
        },
        { parent: this }
      );

      this.hostedZoneId = hostedZone.zoneId;
      this.nameServers = hostedZone.nameServers;
      this.recordName = record.name;

      this.registerOutputs({
        hostedZoneId: this.hostedZoneId,
        nameServers: this.nameServers,
        recordName: this.recordName,
      });
    } else {
      this.registerOutputs({});
    }
  }
}

File: lib/src/main/java/app/DnsStack.java
package app;

import java.util.Map;

import com.pulumi.Context;
import com.pulumi.aws.route53.Record;
import com.pulumi.aws.route53.RecordArgs;
import com.pulumi.aws.route53.Zone;
import com.pulumi.aws.route53.ZoneArgs;
import com.pulumi.aws.route53.inputs.RecordAliasArgs;
import com.pulumi.aws.route53.inputs.RecordGeolocationRoutingPolicyArgs;

/**
 * DNS stack for Route 53 with geolocation routing.
 */
public class DnsStack {
    private final Zone hostedZone;

    public DnsStack(final Context ctx, final CdnStack cdn) {
        // Create Route 53 hosted zone
        this.hostedZone = new Zone("news-portal-zone",
            ZoneArgs.builder()
                .name("newsportal-pr3916.iac-test.com")
                .comment("Hosted zone for news portal")
                .tags(Map.of(
                    "Name", "NewsPortalZone",
                    "Environment", "production"))
                .build());

        // Default record for all locations (with default geolocation)
        var defaultRecord = new Record("news-portal-default",
            RecordArgs.builder()
                .zoneId(hostedZone.zoneId())
                .name("www")
                .type("A")
                .setIdentifier("Default")
                .aliases(
                    RecordAliasArgs.builder()
                        .name(cdn.getDistribution().domainName())
                        .zoneId(cdn.getDistribution().hostedZoneId())
                        .evaluateTargetHealth(false)
                        .build())
                .geolocationRoutingPolicies(
                    RecordGeolocationRoutingPolicyArgs.builder()
                        // Route53 default geolocation catch-all
                        .country("*")
                        .build())
                .build());

        // North America geolocation record
        var northAmericaRecord = new Record("news-portal-na",
            RecordArgs.builder()
                .zoneId(hostedZone.zoneId())
                .name("www")
                .type("A")
                .setIdentifier("NorthAmerica")
                .aliases(
                    RecordAliasArgs.builder()
                        .name(cdn.getDistribution().domainName())
                        .zoneId(cdn.getDistribution().hostedZoneId())
                        .evaluateTargetHealth(true)
                        .build())
                .geolocationRoutingPolicies(
                    RecordGeolocationRoutingPolicyArgs.builder()
                        .continent("NA")
                        .build())
                .build());

        // Europe geolocation record
        var europeRecord = new Record("news-portal-eu",
            RecordArgs.builder()
                .zoneId(hostedZone.zoneId())
                .name("www")
                .type("A")
                .setIdentifier("Europe")
                .aliases(
                    RecordAliasArgs.builder()
                        .name(cdn.getDistribution().domainName())
                        .zoneId(cdn.getDistribution().hostedZoneId())
                        .evaluateTargetHealth(true)
                        .build())
                .geolocationRoutingPolicies(
                    RecordGeolocationRoutingPolicyArgs.builder()
                        .continent("EU")
                        .build())
                .build());

        // Asia geolocation record
        var asiaRecord = new Record("news-portal-as",
            RecordArgs.builder()
                .zoneId(hostedZone.zoneId())
                .name("www")
                .type("A")
                .setIdentifier("Asia")
                .aliases(
                    RecordAliasArgs.builder()
                        .name(cdn.getDistribution().domainName())
                        .zoneId(cdn.getDistribution().hostedZoneId())
                        .evaluateTargetHealth(true)
                        .build())
                .geolocationRoutingPolicies(
                    RecordGeolocationRoutingPolicyArgs.builder()
                        .continent("AS")
                        .build())
                .build());
    }

    public Zone getHostedZone() {
        return hostedZone;
    }
}
